Port files_versions to vue

CarlSchwan opened the pull request:
Simplify code and make it use our standard components, will be helpful when adding more features in the future

After

![Screenshot from 2022-11-09 13-12-16](https://user-images.githubusercontent.com/6653109/200827458-d4526f8c-6ed3-48b5-aa58-8b3983ec573e.png)


![image](https://user-images.githubusercontent.com/23653902/197771479-159a8203-e53a-4421-a548-a5fc71f18eb3.png)

CarlSchwan committed:
Port files_versions to vue

Simplify code and make it use our standard components

Signed-off-by: Carl Schwan <carl@carlschwan.eu>

PVince81 commented:
is the border shadow around the thumbnail on purpose ? are we using this style elsewhere ?

is it on purpose that the row buttons are more blue-ish than the ones in the header ?

PVince81 commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,255 @@
+						{{ version.size }}
+					</div>
+				</div>
+				<NcButton v-tooltip="t('files_version', 'Download this version')"
```
files_versions (missing s)

PVince81 commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,255 @@
+				<NcButton v-tooltip="t('files_version', 'Restore this version')"
+					type="secondary"
+					class="restore-button"
+					:aria-label="t('files_version', 'Restore this version')"
```
to make it unique, include the name / datetime in the aria-label + tooltip

remember: a11y tools can make you "list all links/buttons" in the web page, so their names should be unique

I wonder if we should even go further and make it: "Restore version $version of file $fileName" as aria-label and tooltip

CarlSchwan replied:
fixed

CarlSchwan commented:
> is the border shadow around the thumbnail on purpose ? are we using this style elsewhere ?

Yeah I added them because otherwise the preview is just looking really weird for any text document (and files versions is primarely used for text documents)

![image](https://user-images.githubusercontent.com/23653902/197524621-c571a1fa-5044-4133-a0f6-7c8db69539d6.png)

Alternative we could add either a more subtle shadow or a simple border 

@jancborchardt @nimishavijay what do you think?

> is it on purpose that the row buttons are more blue-ish than the ones in the header ?

It's using secondary action type and not the ternary action type, not sure what is the appropriate here as we don't have any guidelines documented somewhere :(

PVince81 commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,255 @@
+/**
+ *
+ */
+function getDavRequest() {
```
so the new webdav lib still doesn't make it easy to create propfind bodies ?

CarlSchwan replied:
yeah see https://github.com/perry-mitchell/webdav-client#custom-properties :(

PVince81 commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,255 @@
+ * @param modificationTime
+ */
+function formatRelativeTime(modificationTime) {
+	const formatter = new Intl.RelativeTimeFormat('en-US', {
```
I think we're using "moment" JS everywhere, why not here ?

or was there a plan to move away ?

even if we don't use moment, such code would then need to be either in the global JS code in core or in a @nextcloud/... lib somewhere

CarlSchwan replied:
Moment homepage is recomanding against using moment https://momentjs.com/docs/#/-project-status/ 

But yeah let's migrate to Intl later

PVince81 commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,255 @@
+ * @param fileInfo
+ */
+function formatVersion(version, fileInfo) {
+	const fileVersion = version.filename.split('/').reverse()[0] // basename
```
`import { basename } from "@nextcloud/paths"`

PVince81 commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,255 @@
+		mimeType: version.mime,
+		size: OC.Util.humanFileSize(version.size),
+		type: version.type,
+		dateTime: Intl.DateTimeFormat('en-US', { dateStyle: 'full' })
```
use "moment"

PVince81 commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,255 @@
+			.format(new Date(version.lastmod)),
+		relativeTime: formatRelativeTime(new Date(version.lastmod)),
+		preview,
+		url: '/remote.php/dav' + version.filename,
```
maybe use "joinPaths" from `@nextcloud/paths` just to be sure of no missing slashes

PVince81 commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,255 @@
+		/**
+		 * Get the existing versions infos
+		 */
+		async getVersions() {
```
would call it fetchVersions() because "get" usually implies a getter that will return values, but you're not returning anything here

PVince81 commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,255 @@
+					`/versions/${getCurrentUser().uid}/versions/${this.fileInfo.id}/${version.fileVersion}`,
+					`/versions/${getCurrentUser().uid}/restore/target`
+				)
+				showSuccess(t('files_versions', 'Old version restored'))
```
maybe remove "Old" because a version might not be specifically old

PVince81 commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,255 @@
+		margin-left: 1rem;
+		align-self: center;
+	}
+	.downlaod-button {
```
typo

skjnldsv commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,222 @@
+<template>
+	<div>
+		<ul>
+			<li v-for="version in versions" class="version">
```
Elements in iteration expect to have 'v-bind:key' directives

artonge replied:
:robot:

skjnldsv commented:
A few eslint issues to fix :)

CarlSchwan committed:
Handle case with empty version list

Signed-off-by: Carl Schwan <carl@carlschwan.eu>
Signed-off-by: Louis Chemineau <louis@chmn.me>

CarlSchwan commented:
> > Yeah I added them because otherwise the preview is just looking really weird for any text document (and files versions is primarely used for text documents)
> 
> @CarlSchwan no problem, that's what we have `border: 1px solid var(--color-border);` for. ;) And remember the border radius too – just like we do for e.g. Mail attachments and the file list. (Is this a component yet?)
> 
> Drop shadow here does not really make sense as nothing is elevated. :) That is, the text linked with the preview would not have a drop shadow either.

changed to a simple border :) and I updated the PR description with the latest visual

artonge commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,232 @@
+		return {
+			fileInfo: null,
+			versions: [],
+			client,
```
Does the client really needs to be in the component's data ?

artonge commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,232 @@
+ * Get WebDAV request body for version list
+ */
+function getDavRequest() {
+	return `<?xml version="1.0"?>
+			<d:propfind xmlns:d="DAV:"
+				xmlns:oc="http://owncloud.org/ns"
+				xmlns:nc="http://nextcloud.org/ns"
+				xmlns:ocs="http://open-collaboration-services.org/ns">
+				<d:prop>
+					<d:getcontentlength />
+					<d:getcontenttype />
+					<d:getlastmodified />
+				</d:prop>
+			</d:propfind>`
+}
+
+/**
+ * Format version
+ */
+function formatVersion(version, fileInfo) {
+	const fileVersion = basename(version.filename)
+
+	const preview = generateUrl('/apps/files_versions/preview?file={file}&version={fileVersion}', {
+		file: joinPaths(fileInfo.path, fileInfo.name),
+		fileVersion,
+	})
+
+	return {
+		displayVersionName: fileVersion,
+		fileName: version.filename,
+		mimeType: version.mime,
+		size: OC.Util.humanFileSize(version.size),
+		type: version.type,
+		dateTime: moment(version.lastmod),
+		relativeTime: moment(version.lastmod).fromNow(),
+		preview,
+		url: joinPaths('/remote.php/dav', version.filename),
+		fileVersion,
+	}
+}
```
I would put those in a different js file.

artonge commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,232 @@
+		 */
+		async restoreVersion(version) {
+			try {
+				console.debug('restoring version', version.url)
```
```suggestion
				logger.debug('restoring version', version.url)
```

artonge replied:
Not sure if `logger.debug` takes more than 1 arg.

artonge commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,232 @@
+				showSuccess(t('files_versions', 'Version restored'))
+				await this.fetchVersions()
+			} catch (exception) {
+				console.error('Could not restore version', exception)
```
```suggestion
				logger.error('Could not restore version', {exception})
```

artonge reviewed (changes requested):
Beside the logger comment, feel free to ignore the others as they are not crucial :)

artonge commented on apps/files_versions/src/views/VersionTab.vue:
```diff
@@ -0,0 +1,232 @@
+<template>
+	<div>
+		<ul>
+			<li v-for="version in versions" class="version" key="version.url">
```
Also, I would have used `NcListItem`
https://nextcloud-vue-components.netlify.app/#/Components/NcListItems?id=nclistitem

artonge replied:
`NcListItem` is a `<li>` inside, so it is almost a drop in replacement

CarlSchwan replied:
I switched to NcListItem :)

jancborchardt commented:
Gave design feedback in the improvement PR at https://github.com/nextcloud/server/pull/34771/#pullrequestreview-1156318450 :)

CarlSchwan commented:
> - [x]  The wording of the menu is much too complex – as per the spec at https://github.com/nextcloud-gmbh/customer-feature-requests/issues/189#issuecomment-1257905430 it can simply be: "Name this version" (or "Edit version name" if name is set), "Download version", "Restore version" and "Delete version".

Done, this is now using the shorter name. I used the longer name previously because it was inline buttons and we needed for accessibility that the buttons are not the same

> - [ ] The icon in the tab header can be the same as the one you used for the emptycontent and the buttons

Unfortunately this is not possible because the tabs are using the old icons with class and not the vue material icons and there is no way yet to use a material icon with the current API :cry:

CarlSchwan committed:
Use NcListItem

Signed-off-by: Carl Schwan <carl@carlschwan.eu>
Signed-off-by: Louis Chemineau <louis@chmn.me>

CarlSchwan committed:
Add support for Material icon in files sidebar

Signed-off-by: Carl Schwan <carl@carlschwan.eu>
Signed-off-by: Louis Chemineau <louis@chmn.me>

CarlSchwan commented:
@jancborchardt took me 3 hours but I managed to make material icons works, and used the opportunity to move comments and files sharing to also use proper vue icons!

![image](https://user-images.githubusercontent.com/23653902/198061739-bcb5789f-51d8-409a-a4e2-d22a52018039.png)

The material icons are bigger than the old icons, but that needs to be fixed in nextcloud-vue (or port everything to use vue icons)

artonge reviewed (approved):
Looks good to me

skjnldsv reviewed (changes requested):
See comment in https://github.com/nextcloud/activity/pull/981/files#r1005821285

artonge committed:
Use svg icons

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge committed:
Extract logic into separate files

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge committed:
Hide version tab for folders

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge commented:
@skjnldsv I move the icon to plain SVG. Can you validate this approach ?
- I add to create a `raw?` `resourceQuery` for webpack to be able to import raw SVG instead of data URLs.
- I also created a `iconSvg` on the `Tab` class.

artonge committed:
Compile assets

Signed-off-by: Louis Chemineau <louis@chmn.me>

skjnldsv commented on apps/files/src/models/Tab.js:
```diff
@@ -81,12 +84,20 @@ export default class Tab {
 		this._id = id
 		this._name = name
 		this._icon = icon
+		this._iconSvg = iconSvg
```
Doesn't seems to be used? `this._iconSvgSanitized` seems to be the one?

artonge replied:
Ah, yes, forgot to clean that up. Thanks

skjnldsv commented on apps/files/src/views/Sidebar.vue:
```diff
@@ -508,5 +513,13 @@ export default {
+		::v-deep svg {
+			width: 20px;
+			height: 20px;
+			fill: var(--color-main-text);
```
```suggestion
			fill: currentColor;
```

skjnldsv replied:
Careful abou tthe width/height too, you could lead to stretched svg :thinking: 
They're not all square iirc :)

artonge replied:
Didn't know that one, tahnks

skjnldsv commented on apps/files_versions/src/files_versions_tab.js:
```diff
@@ -0,0 +1,69 @@
+const View = Vue.extend(VersionTab)
+let TabInstance = null
+
+window.addEventListener('DOMContentLoaded', function () {
```
```suggestion
window.addEventListener('DOMContentLoaded', function() {
```

skjnldsv commented:
@artonge added comments :)
But we're close to completion! :rocket:

artonge committed:
Lint and remove file_versions tests

Signed-off-by: Louis Chemineau <louis@chmn.me>

artonge commented on apps/comments/src/comments-tab.js:
```diff
@@ -20,12 +20,14 @@
  *
  */
 
+import MessageReplyText from '@mdi/svg/svg/message-reply-text.svg?raw'
```
Disabled eslint for those errors, no idea why it is failing.

artonge commented:
Rebased. Will merge when CI is green
